src/Helpers/Hooks.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace Give\Helpers;
+
+use BadMethodCallException;
+
+class Hooks {
+	/**
+	 * A function which extends the WordPress add_action method to handle the instantiation of a class
+	 * once the action is fired. This prevents the need to instantiate a class before adding it to hook.
+	 *
+	 * @since 2.8.0
+	 *
+	 * @param string $tag
+	 * @param string $class
+	 * @param string $method
+	 * @param int    $priority
+	 * @param int    $acceptedArgs
+	 *
+	 * @return void
+	 */
+	public static function addAction( $tag, $class, $method = '__invoke', $priority = 10, $acceptedArgs = 1 ) {
+		add_action(
+			$tag,
+			static function () use ( $tag, $class, $method ) {
+				// Provide a way of disabling the hook
+				if ( apply_filters( "give_disable_hook-{$tag}", false ) || apply_filters( "give_disable_hook-{$tag}:{$class}@{$method}", false ) ) {
```

Comment:
Would it be more object oriented to expose a `removeAction( $tag, $class, $method )` method, as opposed to an inline filter?

Reply:
@kjohnson Yeah, I thought about this. I have a couple reservations.
1. Folks are used to the hook functions. I felt like documenting a special hook was perhaps simpler than telling them how to pull in our Hooks class and call a method. But I could be convinced others.
2. The bigger challenge is that for this to work we'd have to register the hooks within the class. The reason being that we're using Closures within the WP hooks, so even we don't have a good way of removing them. As such, we'd just have to prevent action within the Closure ourselves. My first thought to do this would be to add a `private $registeredHooks` class property, and the following methods:

```php
	public function removeHook( $tag, $class, $method ) {
		if ( ! $this->isHookRegistered( $tag, $class, $method ) ) {
			return;
		}

		$index = array_search( "$class@method", $this->registeredHooks[ $tag ], true );

		unset( $this->registeredHooks[ $tag ][ $index ] );
	}

	private function registerHook( $tag, $class, $method ) {
		if ( ! isset( $this->registeredHooks[ $tag ] ) ) {
			$this->registeredHooks[ $tag ] = [];
		}

		$this->registeredHooks[ $tag ] = "$class@$method";
	}

	private function isHookRegistered( $tag, $class, $method ) {
		return isset( $this->registeredHooks[ $tag ] ) && in_array( "$class@method", $this->registeredHooks[ $tag ], true );
	}

```

We'd then use these to register the hook, check if the hook is still registered (if not, do not instantiate and fire), and remove hooks.

Thoughts?

Reply:
Your reservations make sense. Also, going back to our previous conversation, the job to be done is memory management and not solving OOP issues in WordPress.

Reply:
Yeah, that's my thought. This route ends up laying our own hooks system on top of WP's. It's still less memory than instantiating, but a bit of a step backwards. I really think it's going to be quite rare for folks to remove our bootstrapping hooks.

So what do you think? Keep the removal system as is?